modules/core/src/main/scala/fs2/kafka/KafkaProducer.scala

```diff
@@ -106,28 +104,27 @@ object KafkaProducer {
     valueSerializer: Serializer[F, V],
     producer: KafkaByteProducer
   )(
-    implicit F: ConcurrentEffect[F]
+    implicit F: Concurrent[F],
+    context: ContextShift[F]
   ): ProducerRecord[K, V] => F[F[(ProducerRecord[K, V], RecordMetadata)]] =
     record =>
       asJavaRecord(keySerializer, valueSerializer, record).flatMap { javaRecord =>
-        Deferred[F, Either[Throwable, (ProducerRecord[K, V], RecordMetadata)]].flatMap { deferred =>
-          F.delay {
-              producer.send(
-                javaRecord,
-                callback { (metadata, exception) =>
-                  val complete =
-                    deferred.complete {
-                      if (exception == null)
-                        Right((record, metadata))
-                      else Left(exception)
-                    }
-
-                  F.runAsync(complete)(_ => IO.unit).unsafeRunSync()
+        F.async { (cb: Either[Throwable, (ProducerRecord[K, V], RecordMetadata)] => Unit) =>
+            producer.send(
+              javaRecord,
+              callback { (metadata, exception) =>
+                cb {
+                  if (exception == null)
+                    Right((record, metadata))
+                  else Left(exception)
                 }
-              )
-            }
-            .as(deferred.get.rethrow)
-        }
+              }
+            )
+            ()
+          }
+          .guarantee(context.shift)
```

Comment:
Is this in the right place? (Don't think I can be bothered to master these CE2 nuances at this stage)(@SystemFw )

Reply:
yeah this is correct